[isoltest] Add support for external sources.

aarlt opened the pull request:
Fixes #10475.

aarlt committed:
[isoltest] Add support for external sources.

axic commented on test/TestCaseReader.cpp:
```diff
@@ -23,12 +23,15 @@
+
 TestCaseReader::TestCaseReader(string const& _filename):
-	m_file(_filename)
+	m_file(_filename), m_filename(_filename)
```
What is the difference between `m_file` and `m_filename`?

axic replied:
Ah, `m_file` is `std::ifstream`, just hard to tell that here in the constructor.

aarlt replied:
Ok, I think I will rename `m_file` to `m_filestream`.

axic commented on test/libsolidity/semanticTests/multiSource/external/external.sol:
```diff
@@ -0,0 +1,11 @@
+// Exclude From Tests
```
I think we may need to ensure that external sources can import too, such as `import "./token.sol";`.

How complicated would that be? Also we do not necessarily want to allow actual test cases in externalsources. The goal would be we include externalsources, and then use them in the importing source to build test cases, etc.

aarlt replied:
Ah thats true!

axic commented on test/libsolidity/semanticTests/multiSource/external/external.sol:
```diff
@@ -0,0 +1,11 @@
+// Exclude From Tests
```
Could we potentially use something in the path for this, like a dot prefix: `semanticTests/multiSource/.external/external.sol`?

aarlt replied:
I did not try this yet, but I would say it should just work. However, I'm not sure, will check that tomorrow.

axic replied:
So lets do underscore (`_`) as the prefix.

axic commented on test/tools/isoltest.cpp:
```diff
@@ -79,9 +79,18 @@ class TestFilter
+		if (regex_match(_name, m_filterExpression))
+		{
+			stringstream stream;
+			ifstream inputStream(_path.string());
```
Wouldn't using `util::readFileAsString` from libsolutil be better?

aarlt replied:
much better!

axic commented on test/TestCaseReader.cpp:
```diff
@@ -136,6 +140,39 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
+					stringstream stream;
+					ifstream externalStream(path.c_str());
+					stream << externalStream.rdbuf();
+					externalSourceContent = stream.str();
```
Wouldn't using `util::readFileAsString` be better?

aarlt commented on test/TestCaseReader.cpp:
```diff
@@ -136,6 +144,54 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
+								fs::path parent(externalSourcePath);
+								parent = parent.parent_path();
+								string importedSourceContent = util::readFileAsString(parent_path.string() + "/" + parent.string() + "/" + import->path());
+								sources[parent.string() + "/" + import->path()] = importedSourceContent;
```
I think it is enough to be able just to point to one solidity file that imports the needed sources. We also could do this recursively, but I think it sufficient to just point to a file that imports all needed external sources.

mijovic replied:
Maybe `(parent / import->path()).string()`?

aarlt replied:
I had that before exactly like this. The problem is that solidity only supports the unix style path separators. If we use  the / operator here, we will have backslashes for windows, that we don't want in this case.

cameel replied:
> I think it is enough to be able just to point to one solidity file that imports the needed sources. We also could do this recursively, but I think it sufficient to just point to a file that imports all needed external sources.

You mean that if, say `a.sol` imports `b.sol` and `b.sol` imports `c.sol`, it's enough to put `a.sol` and `b.sol` in `sources` and the `CompilerStack` will just find the sources on disk? If so, I think this might not work. Path resolution in `CompilerStack` is quite different than here.

> I had that before exactly like this. The problem is that solidity only supports the unix style path separators. If we use the / operator here, we will have backslashes for windows, that we don't want in this case.

If you do `(parent / import->path()).generic_string()`, you will get UNIX backslashes. This is actually how `CompilerStack` does it too.

aarlt replied:
> You mean that if, say `a.sol` imports `b.sol` and `b.sol` imports `c.sol`, it's enough to put `a.sol` and `b.sol` in `sources` and the `CompilerStack` will just find the sources on disk? If so, I think this might not work. Path resolution in `CompilerStack` is quite different than here.
> 

I will add some tests to verify the behaviour. But yes, I thought it could work like this. Let me check.

> If you do `(parent / import->path()).generic_string()`, you will get UNIX backslashes. This is actually how `CompilerStack` does it too.

Ah nice!

cameel replied:
> I will add some tests to verify the behaviour. But yes, I thought it could work like this. Let me check.

I mean, it would work in the simplest case where you have the files in a single directory, importing each other and the working directory is that test directory. It will break when you start using relative imports or subdirectories. And the working directory is usually the repo root so that will also interfere.

cameel replied:
Actually, I have an idea. You could create a custom file loader (the default is [FileReader::readFile()](https://github.com/ethereum/solidity/blob/develop/libsolidity/interface/FileReader.cpp#L45-L94)) that looks for files in the test directory instead of in the current working directory. The loader is basically just lambda and you could ignore all the complications of base path.

Or you could just reuse `FileReader` and only set a custom base path which would be even simpler and you would avoid duplicating the logic.

aarlt replied:
I'm wondering whether this is too complicated. I think complex import situations could be solved by defining a remapping.

cameel replied:
Well, reusing `FileReader` class for this would just require creating a new instance and passing it to `CompilerStack` when it's created. So not that complicated. I think that designating the test directory as the working directory actually does make a lot of sense. It would probably make nested imports "just work" as you originally intended.

But well, we could also just ignore these nested imports in this PR and deal with them separately.

cameel replied:
Actually, forget about reusing `FileReader`. I realized that supporting relative imports and basically any kind of imports is easy as long as we just stop at the virtual filesystem level and never go to the actual filesystem. So please disregard my suggestion above and just look at:
- https://github.com/ethereum/solidity/pull/11110#discussion_r617047235
- https://github.com/ethereum/solidity/pull/11110#issuecomment-823648499

aarlt commented on test/boostTest.cpp:
```diff
@@ -86,14 +86,15 @@ int registerTests(
+		{
+			bool addTest = true;
+			for (auto const& element: entry.path())
+				if (boost::starts_with(element.string(), "_"))
```
Not really sure about this. E.g. if someone checks out the whole solidity  repo in a directory that is placed in a directory starting `_`, all tests would be ignored. I could somehow only check the directories starting from something like `test/libsolidity/semanticTests/`. I guess that would be cleaner.

mijovic replied:
This is interesting. What about maybe adding a comment as a first line in a file that states this is external source file, and then check based on that?

aarlt replied:
Initially I also added a special comment for external sources. The problem is that we need to change the external sources, if we want to use them here. So @axic suggested to use `_` as a prefix. I think thats a nice solution and we don't need to change any external sources, before we can use them in our test framework.

mijovic replied:
Ok, sounds good

mijovic replied:
Maybe we should actually implement like you suggested: check the directories starting from something like `test/libsolidity/semanticTests/`?

aarlt replied:
Yeah, I think that would be better.

aarlt replied:
I add a simple check for `test/libsolidity/semanticTests/`.

aarlt commented on test/libsolidity/ABIEncoderTests.cpp:
```diff
@@ -779,7 +779,7 @@ BOOST_AUTO_TEST_CASE(struct_in_constructor_data_short)
 	NEW_ENCODER(
 		BOOST_CHECK(
-			compileAndRunWithoutCheck({{"", sourceCode}}, 0, "C", encodeArgs(0x20, 0x60, 0x03, 0x80, 0x00)).empty()
+			compileAndRunWithoutCheck({{"", sourceCode}}, nullopt, 0, "C", encodeArgs(0x20, 0x60, 0x03, 0x80, 0x00)).empty()
```
I think I will move that "file" parameter somewhere to the end and set a default value, with this these tests don't need to be changed.

aarlt commented on test/libsolidity/semanticTests/multiSource/_external/import.sol:
```diff
@@ -0,0 +1,2 @@
+import "external.sol";
+import "otherExternal.sol";
```
I think it is ok to support only the import of subsequent external files like this. In such file we just list all the external dependencies that need to be imported. I could also try to resolve all imports recursively, but I think that this is not really needed. What do you think?

mijovic replied:
I think it is fine like this for now. If needed, we can resolve imports recursively in another PR.

aarlt commented on test/TestCaseReader.cpp:
```diff
@@ -136,6 +143,54 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
+								fs::path parent(externalSourcePath);
+								parent = parent.parent_path();
+								string importedSourceContent = util::readFileAsString(parent_path.string() + "/" + parent.string() + "/" + import->path());
+								sources[parent.string() + "/" + import->path()] = importedSourceContent;
```
I guess the usage of that path delimiter is not really portable. Should be changed.

aarlt commented on test/TestCaseReader.cpp:
```diff
@@ -136,6 +143,54 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
+							{
+								fs::path parent(externalSourcePath);
+								parent = parent.parent_path();
+								string importedSourceContent = util::readFileAsString(parent_path.string() + "/" + parent.string() + "/" + import->path());
```
portability..

aarlt commented on docs/grammar/Solidity.g4:
```diff
@@ -39,7 +39,7 @@ importAliases: symbol=identifier (As alias=identifier)?;
 /**
  * Path of a file to be imported.
  */
-path: NonEmptyStringLiteral;
```
I was not able to understand the underlying problem here. Even if I changed `NonEmptyStringLiteral` to exactly the same rule as defined by `StringLiteral`, the rule was not matching.

aarlt replied:
Maybe @ekpyron could take a look. The main problem with the "old" grammar was that it did not support the `import` statements, because that `path` rule did never match (even with an non empty string literal, e.g. `import "asd";`).

aarlt replied:
@ekpyron ah this problem was only related to the rule order.

aarlt replied:
Ok, turned out that just changing the rule order created a lot of other parsing problems. I now just removed `NonEmptyStringLiteral` and use for `path` just `StringLiteral`. The problem is that now `syntaxTests/parsing/import_empty.sol` would fail with `test_antlr_grammar.sh`. So I'm now just ignoring this test in `test_antlr_grammar.sh`.

ekpyron replied:
I just created https://github.com/ethereum/solidity/pull/11229 to properly deal with all of that. If you undo all grammar changes here and just rebase on top of that PR, everything should just work (and it should even test the grammar against the external projects and against external sources, both of which makes a lot of sense I'd say).

aarlt replied:
Great!

aarlt commented on docs/grammar/Solidity.g4:
```diff
@@ -1,7 +1,7 @@
  * Solidity is a statically typed, contract-oriented, high-level language for implementing smart contracts on the Ethereum platform.
  */
-grammar Solidity;
+parser grammar SolidityParser;
```
I was not able to debug rules without explicitly saying that this is a parser grammar. However, doing this here has also implications on how `test_antlr_grammar.sh` is executed. Somehow the `SolidityParser.class` is not found anymore, we just name that grammar `Solidity`, so I needed to change that to `SolidityParser`. For `scripts/doc.sh` needs to have `docs/grammar/Solidity.g4` present. To be able to review that easier, I only changed `Solidity.g4` and created a symbolic link `SolidityParser.g4`.

stackenbotten commented:
There was an error when running `chk_coding_style` for commit `a8783a25fd16ba765874a7fc56495aa300a43679`:
```
Error: Trailing whitespace found:
test/libsolidity/semanticTests/externalContracts/openzeppelin/_contracts/mocks/DummyImplementation.sol:6:  function version() public pure virtual returns (string memory); 

```
Please check that your changes are working as intended.

aarlt commented on test/libsolidity/semanticTests/externalContracts/openzeppelin/Ownable.sol:
```diff
@@ -0,0 +1,25 @@
+// compileViaYul: also
+// ----
+// store(uint256): 123
+// retrieve() -> 123
```
Maybe we should add an isoltest builtin that allows changing the current used account.

aarlt commented:
Oha, I guess we also want to prevent running the style & spell-checker on external sources.

mijovic commented on libsolidity/interface/CompilerStack.cpp:
```diff
@@ -631,16 +631,17 @@ vector<string> CompilerStack::contractNames() const
 }
 
-string const CompilerStack::lastContractName() const
+string const CompilerStack::lastContractName(std::optional<std::string> const& _sourceName) const
```
At least for string you can get rid off `std`
```suggestion
string const CompilerStack::lastContractName(optional<string> const& _sourceName) const
```

mijovic commented on test/TestCaseReader.cpp:
```diff
@@ -136,6 +143,54 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
+				{
+					ErrorList errorList;
+					ErrorReporter errorReporter(errorList);
+					Parser parser(errorReporter, langutil::EVMVersion::berlin());
```
Is there a reason why `berlin()`, why not just `langutil::EVMVersion{}`?

mijovic commented on test/libsolidity/SolidityExecutionFramework.cpp:
```diff
@@ -36,6 +36,7 @@ using namespace std;
 
 bytes SolidityExecutionFramework::multiSourceCompileContract(
 	map<string, string> const& _sourceCode,
+	std::optional<std::string> const& _sourceName,
```
Removing `std`?

mijovic reviewed (commented):
Only few small things

mijovic commented on test/libsolidity/semanticTests/externalContracts/openzeppelin/_contracts/metatx/README.adoc:
```diff
@@ -0,0 +1,12 @@
+= Meta Transactions
```
Do we need this file?

aarlt replied:
I just copied all files. Maybe we don't need all files, but I think its easier to just copy everything.

mijovic commented on test/boostTest.cpp:
```diff
@@ -86,14 +86,15 @@ int registerTests(
-					_labels,
-					_testCaseCreator
-				);
+					*sub_suite, _basepath, _path / entry.path().filename(), _enforceViaYul, _labels, _testCaseCreator);
```
Why merging into one line? I think it doesn't align to code style as well

mijovic commented on test/libsolidity/SolidityExecutionFramework.cpp:
```diff
@@ -126,6 +127,7 @@ bytes SolidityExecutionFramework::compileContract(
 {
 	return multiSourceCompileContract(
 		{{"", _sourceCode}},
+		std::nullopt,
```
```suggestion
	nullopt,
```

mijovic commented:
Do we really need all these files from openzeppelin, I am mostly thinking about docs?

ekpyron commented:
In case the comment way up there might get lost otherwise, I leave this as big fat comment here once more: https://github.com/ethereum/solidity/pull/11229 contains more proper antlr grammar fixes that are already completely compatible with this PR, so just undoing all antlr changes here and rebasing on that should fix all grammar-related issues.

mijovic commented on test/libsolidity/SolidityExecutionFramework.cpp:
```diff
@@ -68,7 +69,7 @@ bytes SolidityExecutionFramework::multiSourceCompileContract(
 		BOOST_ERROR("Compiling contract failed");
 	}
-	std::string contractName(_contractName.empty() ? m_compiler.lastContractName() : _contractName);
+	std::string contractName(_contractName.empty() ? m_compiler.lastContractName(_sourceName) : _contractName);
```
```suggestion
	string contractName(_contractName.empty() ? m_compiler.lastContractName(_sourceName) : _contractName);
```

mijovic commented on test/boostTest.cpp:
```diff
@@ -86,14 +86,20 @@ int registerTests(
+			for (auto const& element: entry.path())
+				if (boost::starts_with(element.string(), "_"))
+					addTest = false;
+			if (addTest && (fs::is_directory(entry.path()) || TestCase::isTestFilename(entry.path().filename())))
```
I think it is hard to read as one line, maybe splitting like this
```suggestion
			if (
				addTest && 
				(fs::is_directory(entry.path()) || TestCase::isTestFilename(entry.path().filename()))
			)
```

mijovic commented on test/TestCaseReader.cpp:
```diff
@@ -137,6 +144,54 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
+					externalSourcePath = boost::trim_copy(externalSourcePath.substr(remappingPos + 1));
+				}
+
+				string externalSourceContent;
```
Can you move it down just before `if` statement where it is being set actually?

mijovic commented on test/TestCaseReader.cpp:
```diff
@@ -137,6 +144,54 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
+				{
+					ErrorList errorList;
+					ErrorReporter errorReporter(errorList);
+					Parser parser(errorReporter, langutil::EVMVersion{});
+					shared_ptr<Scanner> scanner = make_shared<Scanner>(CharStream(externalSourceContent, externalSourceName));
+					ASTPointer<SourceUnit> sourceUnit = parser.parse(scanner);
```
```suggestion
					ASTPointer<SourceUnit> sourceUnit = Parser(errorReporter, langutil::EVMVersion{}).parse(
						make_shared<Scanner>(CharStream(externalSourceContent, externalSourceName))
					);
```

mijovic commented:
Can you also add a semantic test that uses one of the external sources from OZ?
For example inherit from `Ownable`?

mijovic commented:
And, linking one more comment to not get lost https://github.com/ethereum/solidity/pull/11110#discussion_r613111651

aarlt commented:
> Can you also add a semantic test that uses one of the external sources from OZ?
> For example inherit from `Ownable`?

@mijovic I added already a semantic test for this: e.g. https://github.com/ethereum/solidity/blob/9345590866753bc81677411d9ef07a33eeef76fe/test/libsolidity/semanticTests/externalContracts/openzeppelin/Ownable.sol`, or what do you mean?

mijovic commented:
> > Can you also add a semantic test that uses one of the external sources from OZ?
> > For example inherit from `Ownable`?
> 
> @mijovic I added already a semantic test for this: e.g. [9345590#diff-a96982cc40d35ca4241edc0e55b4619591f70059538462f8fa9c3c043b179af1](https://github.com/ethereum/solidity/commit/9345590866753bc81677411d9ef07a33eeef76fe#diff-a96982cc40d35ca4241edc0e55b4619591f70059538462f8fa9c3c043b179af1), or what do you mean?

Ah, sorry, somehow lost it in list of files. I kind of seen it when did a first review.

cameel commented on .circleci/config.yml:
```diff
@@ -299,7 +299,7 @@ jobs:
       - run:
           name: Check spelling
-          command: ~/.local/bin/codespell -S "*.enc,.git,Dockerfile*" -I ./scripts/codespell_whitelist.txt
+          command: ~/.local/bin/codespell --check-filenames -S "*.enc,.git,Dockerfile*,_*" -I ./scripts/codespell_whitelist.txt
```
Which files is this meant to exclude? I don't see any files starting with `_` in the PR.

On the other hand we have `docs/_templates/` and `docs/_static/`. They currently do not contain too much text but still, I don't see a reason to exclude them.

aarlt replied:
The main problem here was that the OZ sources contained a lot of spelling errors and/or unknown words. So I thought it made sense to exclude external sources from being spell-checked.

cameel commented on scripts/check_style.sh:
```diff
@@ -16,6 +16,13 @@ EXCLUDE_FILES=(
+# where its name is starting with '_' (e.g. for external sources used in semantic tests).
+for EXTERNAL_SOURCE in $(find test/libsolidity/semanticTests | grep "/_")
+do
+  EXCLUDE_FILES+=( "${EXTERNAL_SOURCE}" )
```
```suggestion
    EXCLUDE_FILES+=("${EXTERNAL_SOURCE}")
```

cameel commented on scripts/check_style.sh:
```diff
@@ -16,6 +16,13 @@ EXCLUDE_FILES=(
+
+# Also exclude files that are placed within a directory,
+# where its name is starting with '_' (e.g. for external sources used in semantic tests).
+for EXTERNAL_SOURCE in $(find test/libsolidity/semanticTests | grep "/_")
```
Why not pass the name directly to `find`? Would this work?

```suggestion
for EXTERNAL_SOURCE in $(find test/libsolidity/semanticTests -name '*/_*')
```

aarlt replied:
I reverted the changes here. If we add OZ or another repository, we can easily add a similar mechanism to exclude such files from being checked. But for this PR it is not needed.

cameel commented on scripts/codespell_whitelist.txt:
```diff
@@ -13,3 +13,4 @@ hist
 otion
 keypair
 ether
+arithmetics
```
I don't see this word used anywhere in this PR and to me it looks like an actual mistake. Are you sure you want to add it?

aarlt replied:
I added ` --check-filenames` to `codespell`, but maybe this should not be done in this PR. I will remove this.

cameel commented:
@aarlt Not sure if this was already discussed (can't find any mention in the comments here) but do we really want to dump all the OZ files directly into the repo? Why not use a git submodule?

Alternatively, we already maintain a clone of the OZ repo in https://github.com/solidity-external-tests/openzeppelin-contracts. Maybe you could just check it out in CI before running the test?

cameel commented on libsolidity/interface/CompilerStack.cpp:
```diff
@@ -695,16 +695,17 @@ vector<string> CompilerStack::contractNames() const
-		for (ASTPointer<ASTNode> const& node: it.second.ast->nodes())
-			if (auto contract = dynamic_cast<ContractDefinition const*>(node.get()))
-				contractName = contract->fullyQualifiedName();
+		if (!_sourceName.has_value() || (_sourceName.has_value() && it.first == _sourceName.value()))
```
```suggestion
		if (!_sourceName.has_value() || it.first == _sourceName.value())
```

cameel commented on test/TestCaseReader.cpp:
```diff
@@ -18,24 +18,30 @@
-	m_file(_filename)
+namespace fs = boost::filesystem;
+
+TestCaseReader::TestCaseReader(string const& _filename): m_filestream(_filename), m_filename(_filename)
```
```suggestion
TestCaseReader::TestCaseReader(string const& _filename):
    m_filestream(_filename),
    m_filename(_filename)
```

cameel commented on test/TestCaseReader.h:
```diff
@@ -62,7 +62,8 @@ class TestCaseReader
 
-	std::ifstream m_file;
+	std::ifstream m_filestream;
+	std::string m_filename;
```
```suggestion
	std::ifstream m_fileStream;
	std::string m_fileName;
```

cameel commented on test/TestCaseReader.cpp:
```diff
@@ -137,6 +144,54 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
+				}
+
+				string externalSourceContent;
+				fs::path parent_path(m_filename);
```
```suggestion
				fs::path testCaseParentDir(m_filename);
```

cameel commented on test/TestCaseReader.cpp:
```diff
@@ -137,6 +144,54 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
+				}
+
+				string externalSourceContent;
+				fs::path parent_path(m_filename);
+				parent_path = parent_path.normalize().remove_filename();
```
`remove_filename()` modifies the path in place. I'd do either
```suggestion
				fs::path parent_path = fs::path(m_filename).normalize();
				parent_path.remove_filename();
```
or
```suggestion
				fs::path parent_path = fs::path(m_filename).normalize().parent_path();
```

By the way, where does `normalize()` come from? I do not see it in the [docs for `boost::filesystem` in Boost 1.75.0](https://www.boost.org/doc/libs/1_75_0/libs/filesystem/doc/reference.html) and I saw comments mentioning that it has been removed from boost. That's weird because the tests pass in this PR.

aarlt replied:
Good catch! Looks like `normalize()` got deprecated. I use `canonical()` now.

cameel commented on test/TestCaseReader.cpp:
```diff
@@ -137,6 +144,54 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
+				if (fs::exists(path))
+					externalSourceContent = util::readFileAsString(path.string());
+				else
+					BOOST_THROW_EXCEPTION(runtime_error(string("External Source '"+ externalSourcePath +"' not found.")));
```
```suggestion
					BOOST_THROW_EXCEPTION(runtime_error("External Source '"s + externalSourcePath + "' not found."));
```

cameel commented on test/TestCaseReader.cpp:
```diff
@@ -137,6 +144,54 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
+				fs::path parent_path(m_filename);
+				parent_path = parent_path.normalize().remove_filename();
+
+				fs::path path = parent_path / fs::path(externalSourcePath);
```
What happens if `externalSourcePath` is absolute?

aarlt replied:
I would say it need to be always relative. I added an assert.

cameel commented on test/TestCaseReader.cpp:
```diff
@@ -137,6 +144,54 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
+				fs::path parent_path(m_filename);
+				parent_path = parent_path.normalize().remove_filename();
+
+				fs::path path = parent_path / fs::path(externalSourcePath);
```
There are so many paths in here that calling it just `path` is very confusing.
```suggestion
				fs::path externalSourceFullPath = parent_path / fs::path(externalSourcePath);
```

cameel commented on test/TestCaseReader.cpp:
```diff
@@ -137,6 +144,54 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
 			}
+			else if (boost::algorithm::starts_with(line, externalSourceDelimiterStart) && boost::algorithm::ends_with(line, sourceDelimiterEnd))
+			{
+				string externalSourcePath = boost::trim_copy(line.substr(
+					externalSourceDelimiterStart.size(),
+					line.size() - sourceDelimiterEnd.size() - externalSourceDelimiterStart.size()
+				));
+				string externalSourceName = externalSourcePath;
+
+				// Does the external source define a remapping?
+				size_t remappingPos = externalSourcePath.find('=');
+				if (remappingPos != string::npos)
+				{
+					externalSourceName = boost::trim_copy(externalSourcePath.substr(0, remappingPos));
+					externalSourcePath = boost::trim_copy(externalSourcePath.substr(remappingPos + 1));
+				}
+
+				string externalSourceContent;
+				fs::path parent_path(m_filename);
+				parent_path = parent_path.normalize().remove_filename();
+
+				fs::path path = parent_path / fs::path(externalSourcePath);
```
How about extracting this part into a helper?

Maybe even adding a test for it?

aarlt replied:
I moved the tests to `test/libsolidity/semanticTests/externalSource`. There are also some related to remapping.

aarlt replied:
Ah, somehow I missed this. What part would you like to extract into an helper? All the lines `149-167`? I added a simple helper that checks whether a test is part of an external source directory (starting with `_`). See https://github.com/ethereum/solidity/pull/11110#discussion_r616746218. But I'm not sure what kind of helper you meant here.

cameel replied:
Yeah, I was thinking that the whole piece above could be a separate helper. It only processes paths so it could be easily tested.

cameel commented on test/TestCaseReader.cpp:
```diff
@@ -137,6 +144,54 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
+					ASTPointer<SourceUnit> sourceUnit = parser.parse(scanner);
+					if (sourceUnit)
+						for (auto const& node: sourceUnit->nodes())
+							if (auto import = dynamic_cast<ImportDirective const*>(node.get()))
```
```suggestion
						for (auto const* import: ASTNode::filteredNodes<ImportDirective>(sourceUnit->nodes()))
```

cameel commented on test/TestCaseReader.cpp:
```diff
@@ -137,6 +144,54 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
+							if (auto import = dynamic_cast<ImportDirective const*>(node.get()))
+							{
+								fs::path parent(externalSourcePath);
+								parent = parent.parent_path();
```
Please give it a more self-explanatory name. There are multiple parent dirs in this function.
```suggestion
								fs::path externalSourceParentDir = fs::path(externalSourcePath).parent_path();
```

cameel commented on test/TestCaseReader.cpp:
```diff
@@ -137,6 +144,54 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
+							{
+								fs::path parent(externalSourcePath);
+								parent = parent.parent_path();
+								string importedSourceContent = util::readFileAsString((parent_path / parent / import->path()).string());
```
Again, what if `parent` or `import->path()` are absolute paths? If absolute paths are not supported, you should have a check against that.

cameel commented on test/TestCaseReader.cpp:
```diff
@@ -137,6 +144,54 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
+							{
+								fs::path parent(externalSourcePath);
+								parent = parent.parent_path();
+								string importedSourceContent = util::readFileAsString((parent_path / parent / import->path()).string());
```
This does not handle relative imports correctly.

This code will handle `import "contract.sol"` and `import "./contract.sol"` the same way but they can be different files. The first one is relative to compiler's working directory while the other is relative to the location of the importing file. See sections on "relative imports" and "base path" in #11263.

cameel replied:
After reviewing the hole PR I think that https://github.com/ethereum/solidity/pull/11110#discussion_r617047235 is a better solution to relative imports. So I'm marking this as resolved and let's focus on that other solution.

cameel commented on test/TestCaseReader.cpp:
```diff
@@ -137,6 +144,54 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
+					Parser parser(errorReporter, langutil::EVMVersion{});
+					shared_ptr<Scanner> scanner = make_shared<Scanner>(CharStream(externalSourceContent, externalSourceName));
+					ASTPointer<SourceUnit> sourceUnit = parser.parse(scanner);
+					if (sourceUnit)
```
What is intended to happen when the file fails parsing? Will it fail at a later stage or are imports just ignored?

aarlt replied:
I added an assert.

cameel replied:
Oh, I thought it might have been intentional. That's why I asked instead of suggesting a change :)

So this is meant to support only semantic tests where the code is compilable, right? No syntax tests?

aarlt replied:
Yes, I think so.

cameel replied:
I think you should remove this condition. It's superfluous now that you have the assert.

cameel commented on test/TestCaseReader.cpp:
```diff
@@ -137,6 +144,54 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
+					ErrorList errorList;
+					ErrorReporter errorReporter(errorList);
+					Parser parser(errorReporter, langutil::EVMVersion{});
+					shared_ptr<Scanner> scanner = make_shared<Scanner>(CharStream(externalSourceContent, externalSourceName));
+					ASTPointer<SourceUnit> sourceUnit = parser.parse(scanner);
```
You are assuming that the external source is an ordinary solidity file. What if it is a multi-source test case as well?

aarlt replied:
Correct, this use-case will not be supported. I think it is ok to just support ordinary solidity files.

cameel replied:
ok. Sounds reasonable.

cameel commented on test/TestCaseReader.cpp:
```diff
@@ -137,6 +144,54 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
+				{
+					ErrorList errorList;
+					ErrorReporter errorReporter(errorList);
+					Parser parser(errorReporter, langutil::EVMVersion{});
```
Not sure how much it matters while parsing, but I think you should respect the value of `--evm-version` parameter here.

aarlt replied:
Yeah I thought the same. Let me see how I can get the version here.

axic commented:
I'd suggest not to import openzeppelin in this PR. And when we do import openzeppelin partially or completely, the commit should say what exact version we're are importing, and the actual semantic tests should make good use of all that code we import.

aarlt commented:
I just noticed that updating expectations using isoltest are not working right now. Basically the source gets fully rewritten by the updating mechanism, where external sources get directly added to the source.

mijovic commented on libsolidity/interface/CompilerStack.cpp:
```diff
@@ -695,16 +695,17 @@ vector<string> CompilerStack::contractNames() const
-		for (ASTPointer<ASTNode> const& node: it.second.ast->nodes())
-			if (auto contract = dynamic_cast<ContractDefinition const*>(node.get()))
-				contractName = contract->fullyQualifiedName();
+		if (!_sourceName.has_value() || it.first == _sourceName.value())
```
More compact and clean maybe?
```suggestion
		if (_sourceName.value_or(it.first) == it.first)
```

mijovic commented on test/TestCaseReader.cpp:
```diff
@@ -137,6 +146,56 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
+					if (sourceUnit)
+						for (auto const* import: ASTNode::filteredNodes<ImportDirective>(sourceUnit->nodes()))
+						{
+							fs::path externalSourceParentDir(externalSource);
+							externalSourceParentDir = externalSourceParentDir.parent_path();
```
Maybe as one line?
```suggestion
							fs::path externalSourceParentDir = fs::path(externalSource).parent_path();
```

mijovic commented on test/TestCaseReader.cpp:
```diff
@@ -137,6 +146,56 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
+							fs::path externalSourceParentDir(externalSource);
+							externalSourceParentDir = externalSourceParentDir.parent_path();
+							solAssert(externalSourceParentDir.is_relative() && fs::path(import->path()).is_relative(), "");
+							string importedSourceContent = util::readFileAsString((testCaseParentDir / externalSourceParentDir / import->path()).string());
```
This line is too long
```suggestion
							string importedSourceContent = util::readFileAsString(
								(testCaseParentDir / externalSourceParentDir / import->path()).string()
							);
```

mijovic commented on test/boostTest.cpp:
```diff
@@ -86,14 +86,30 @@ int registerTests(
+			for (auto const& element: entry.path())
+			{
+				parent_path = parent_path / element;
+				if (boost::ends_with(canonical(parent_path).generic_string(), "test/libsolidity/semanticTests"))
```
Are you sure that this works properly? I think path will not end with "test/libsolidity/semanticTests" as you already added `element` to `parent_path`

mijovic commented on test/tools/isoltest.cpp:
```diff
@@ -79,9 +80,23 @@ class TestFilter
+			for (auto const& element: _path)
+			{
+				parent_path = parent_path / element;
+				if (boost::ends_with(canonical(parent_path).generic_string(), "test/libsolidity/semanticTests"))
```
Maybe we can add a small helper for this as it might change sometime in future, so we do not need to maintain same code in two different places...

mijovic replied:
Actually whole logic for including test should be helper. Can you refactor it please?

cameel replied:
Yeah. That's what I was suggesting in https://github.com/ethereum/solidity/pull/11110#discussion_r614766304

aarlt replied:
Ok, I added a helper to `test/Common.h`. The helper is `bool isValidTestPath(boost::filesystem::path const& _path);` not sure thats a good name for that.

mijovic replied:
Maybe to be more precise and say `isValidSemanticTestPath`?

mijovic commented on test/TestCaseReader.cpp:
```diff
@@ -137,6 +146,57 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
+					externalSource = boost::trim_copy(externalSource.substr(remappingPos + 1));
+				}
+
+				fs::path testCaseParentDir(m_fileName);
+				testCaseParentDir = canonical(testCaseParentDir).parent_path();
```
I find it more readable like this, but not mandatory to update
```suggestion
				fs::path testCaseParentDir = canonical(fs::path{m_fileName}).parent_path();
```

mijovic commented on test/tools/isoltest.cpp:
```diff
@@ -79,9 +80,11 @@ class TestFilter
+	bool matches(fs::path const& _path, string const& _name) const
 	{
-		return regex_match(_name, m_filterExpression);
+		if (regex_match(_name, m_filterExpression))
+			return solidity::test::isValidTestPath(_path);
+		return false;
```
```suggestion
		return regex_match(_name, m_filterExpression) && solidity::test::isValidTestPath(_path);
```

mijovic commented on test/boostTest.cpp:
```diff
@@ -86,14 +86,19 @@ int registerTests(
 			fs::directory_iterator()
 		))
-			if (fs::is_directory(entry.path()) || TestCase::isTestFilename(entry.path().filename()))
+		{
```
You can remove brackets as it is single statement inside if now

cameel commented on .circleci/config.yml:
```diff
@@ -299,7 +299,7 @@ jobs:
       - run:
           name: Check spelling
-          command: ~/.local/bin/codespell -S "*.enc,.git,Dockerfile*" -I ./scripts/codespell_whitelist.txt
+          command: ~/.local/bin/codespell -S "*.enc,.git,Dockerfile*,_*" -I ./scripts/codespell_whitelist.txt
```
So if we are removing OZ from the PR, I think we should remove this too.

cameel commented on libsolidity/interface/CompilerStack.cpp:
```diff
@@ -695,16 +695,17 @@ vector<string> CompilerStack::contractNames() const
-			if (auto contract = dynamic_cast<ContractDefinition const*>(node.get()))
-				contractName = contract->fullyQualifiedName();
+		if (_sourceName.value_or(it.first) == it.first)
+			for (ASTPointer<ASTNode> const& node: it.second.ast->nodes())
+				if (auto contract = dynamic_cast<ContractDefinition const*>(node.get()))
```
```suggestion
			for (auto const* contract: ASTNode::filteredNodes<ContractDefinition>(it.second.ast->nodes()))
```

cameel commented on test/Common.cpp:
```diff
@@ -222,4 +223,19 @@ void CommonOptions::setSingleton(std::unique_ptr<CommonOptions const>&& _instanc
+	for (auto const& element: _path)
+	{
+		parent_path = parent_path / element;
+		if (boost::ends_with(canonical(parent_path).generic_string(), "test/libsolidity/semanticTests"))
```
I think you should ensure that there's either `/` before `test/` or it's the beginning of the string.

cameel commented on test/Common.cpp:
```diff
@@ -222,4 +223,19 @@ void CommonOptions::setSingleton(std::unique_ptr<CommonOptions const>&& _instanc
 
 std::unique_ptr<CommonOptions const> CommonOptions::m_singleton = nullptr;
 
+bool isValidSemanticTestPath(boost::filesystem::path const& _path)
+{
+	bool insideSemanticTests = false;
+	fs::path parent_path;
+	for (auto const& element: _path)
+	{
+		parent_path = parent_path / element;
```
Some naming suggestions to make code more self-explanatory (and use camer case);
```suggestion
bool isValidSemanticTestPath(boost::filesystem::path const& _testPath)
{
	bool insideSemanticTests = false;
	fs::path testPathPrefix;
	for (auto const& segment: _testPath)
	{
		testPathPrefix = testPathPrefix / segment;
```

cameel replied:
I wonder if `/=` works with `boost::filesystem::path`?

```suggestion
bool isValidSemanticTestPath(boost::filesystem::path const& _path)
{
	bool insideSemanticTests = false;
	fs::path parent_path;
	for (auto const& element: _path)
	{
		parent_path /= element;
```

cameel commented on test/Common.h:
```diff
@@ -83,4 +83,6 @@ struct CommonOptions: boost::noncopyable
 	static std::unique_ptr<CommonOptions const> m_singleton;
 };
 
+bool isValidSemanticTestPath(boost::filesystem::path const& _testPath);
```
This function needs some documentation:

```suggestion
/// @return true if it is ok to treat the file located under the specified path as a semantic test.
/// I.e. if the test is located in the semantic test directory and is not excluded due to being a part of extenal sources.
/// Note: @p _testPath can be relative but must include at least the `/test/libsolidity/semanticTests/` part.
bool isValidSemanticTestPath(boost::filesystem::path const& _testPath);
```

cameel commented on test/TestCaseReader.cpp:
```diff
@@ -137,6 +146,56 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
 			}
+			else if (boost::algorithm::starts_with(line, externalSourceDelimiterStart) && boost::algorithm::ends_with(line, sourceDelimiterEnd))
+			{
+				string externalSource = boost::trim_copy(line.substr(
```
```suggestion
				string externalSourceTarget = boost::trim_copy(line.substr(
```

cameel commented on test/TestCaseReader.cpp:
```diff
@@ -137,6 +146,56 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
+				{
+					externalSourceName = boost::trim_copy(externalSource.substr(0, remappingPos));
+					externalSource = boost::trim_copy(externalSource.substr(remappingPos + 1));
+				}
```
Putting the assignment in the `else` branch would make it much clearer that there are two alternatives here: either it's a remapping and name and target are different or it's not and they're the same.
```suggestion
				// Does the external source define a remapping?
				size_t remappingPos = externalSource.find('=');
				if (remappingPos != string::npos)
				{
					externalSourceName = boost::trim_copy(externalSource.substr(0, remappingPos));
					externalSource = boost::trim_copy(externalSource.substr(remappingPos + 1));
				}
				else
					string externalSourceName = externalSource;
```

cameel commented on test/TestCaseReader.cpp:
```diff
@@ -137,6 +146,56 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
+					externalSource = boost::trim_copy(externalSource.substr(remappingPos + 1));
+				}
+
+				fs::path testCaseParentDir = canonical(fs::path{m_fileName}).parent_path();
```
I think there is an implicit conversion from string:
```suggestion
				fs::path testCaseParentDir = canonical(m_fileName).parent_path();
```

cameel replied:
And now `m_fileName` is `fs::path`.

cameel commented on test/TestCaseReader.h:
```diff
@@ -62,7 +62,8 @@ class TestCaseReader
 
-	std::ifstream m_file;
+	std::ifstream m_fileStream;
+	std::string m_fileName;
```
```suggestion
	boost::filesystem::path m_fileName;
```

cameel commented on test/TestCaseReader.cpp:
```diff
@@ -137,6 +146,56 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
+				}
+
+				fs::path testCaseParentDir = canonical(fs::path{m_fileName}).parent_path();
+				fs::path externalSourcePath(externalSource);
```
This variable seems unnecessary. You can just use `externalSource` directly and have one less name to remember. You're using it that way already in `!fs::path(externalSource).is_relative()` below.

Altenatively, you could make externalSource a `boost::filesystem::path` from the start. You would just have to use `externalSourceName` in place of `externalSource` in `string externalSource = boost::trim_copy(...` at the beginning.

cameel commented on test/TestCaseReader.cpp:
```diff
@@ -137,6 +146,56 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
+				fs::path testCaseParentDir = canonical(fs::path{m_fileName}).parent_path();
+				fs::path externalSourcePath(externalSource);
+				if (!fs::path(externalSource).is_relative())
+					BOOST_THROW_EXCEPTION(runtime_error(string("External Source need to be relative.")));
```
```suggestion
					BOOST_THROW_EXCEPTION(runtime_error("External Source paths need to be relative to the location of the test case."s));
```

cameel commented on test/TestCaseReader.cpp:
```diff
@@ -137,6 +146,56 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
+				if (fs::exists(externalSourceFullPath))
+					externalSourceContent = util::readFileAsString(externalSourceFullPath.string());
+				else
+					BOOST_THROW_EXCEPTION(runtime_error("External Source '" + externalSourcePath.string() + "' not found."));
```
```suggestion
				if (!fs::exists(externalSourceFullPath))
					BOOST_THROW_EXCEPTION(runtime_error("External Source '" + externalSourcePath.string() + "' not found."));
				string externalSourceContent = util::readFileAsString(externalSourceFullPath.string());
```

cameel commented on test/TestCaseReader.cpp:
```diff
@@ -137,6 +146,56 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
+
+				string externalSourceContent;
+				if (fs::exists(externalSourceFullPath))
+					externalSourceContent = util::readFileAsString(externalSourceFullPath.string());
```
Why not add `using solidity::util` at the beginning of the file?
```suggestion
					externalSourceContent = readFileAsString(externalSourceFullPath.string());
```

cameel commented on test/TestCaseReader.cpp:
```diff
@@ -137,6 +146,56 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
+					ErrorReporter errorReporter(errorList);
+					shared_ptr<Scanner> scanner = make_shared<Scanner>(CharStream(externalSourceContent, externalSourceName));
+					ASTPointer<SourceUnit> sourceUnit = Parser(errorReporter, evmVersion).parse(scanner);
+					solAssert(sourceUnit != nullptr, "");
```
I think I'd use `BOOST_THROW_EXCEPTION` here with an error message containing the path to the file that has syntax errors. I think that having a broken test will be a pretty common situation when writing or modifying tests. We should make it easy to find and fix the broken file.

aarlt replied:
However, I removed the parsing part completely now.

cameel commented on test/TestCaseReader.cpp:
```diff
@@ -137,6 +146,56 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
+				else
+					BOOST_THROW_EXCEPTION(runtime_error("External Source '" + externalSourcePath.string() + "' not found."));
+
+				if (!externalSourceName.empty())
+				{
```
I think we should just report an error (using `BOOST_THROW_EXCEPTION`) when the name is empty. Or is there are valid situation where this is useful?

aarlt replied:
I put this as an assert.

cameel replied:
This can actually happen in normal circumstances. I.e. if you just do `==== ExternalSources: ====` and forget the name. I think that a normal error is more appropriate in this case.

cameel commented on test/TestCaseReader.cpp:
```diff
@@ -137,6 +146,56 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
+						for (auto const* import: ASTNode::filteredNodes<ImportDirective>(sourceUnit->nodes()))
+						{
+							fs::path externalSourceParentDir = fs::path(externalSource).parent_path();
+							solAssert(externalSourceParentDir.is_relative() && fs::path(import->path()).is_relative(), "");
```
I think that paths used in imports do not have to be restricted like this. If they are absolute, they just have to be remapped so accepting anything is fine.
```suggestion
							solAssert(externalSourceParentDir.is_relative(), "");
```

cameel commented on test/boostTest.cpp:
```diff
@@ -86,7 +86,10 @@ int registerTests(
+			if (
+					solidity::test::isValidSemanticTestPath(entry) &&
+					(fs::is_directory(entry.path()) || TestCase::isTestFilename(entry.path().filename()))
+			)
```
```suggestion
			if (
				solidity::test::isValidSemanticTestPath(entry) &&
				(fs::is_directory(entry.path()) || TestCase::isTestFilename(entry.path().filename()))
			)
```

cameel commented on test/libsolidity/semanticTests/externalSource/_external/external.sol:
```diff
@@ -0,0 +1,5 @@
+  function foo() public pure returns (uint) {
+    return 1234;
+  }
+}
```
We should follow the convention we recommend to users, i.e. 4-space indents in Solidity code:
```suggestion
contract External {
    function foo() public pure returns (uint) {
        return 1234;
  }
}
```

axic commented on libsolidity/interface/CompilerStack.cpp:
```diff
@@ -695,15 +695,15 @@ vector<string> CompilerStack::contractNames() const
 	for (auto const& it: m_sources)
-		for (ASTPointer<ASTNode> const& node: it.second.ast->nodes())
-			if (auto contract = dynamic_cast<ContractDefinition const*>(node.get()))
+		if (_sourceName.value_or(it.first) == it.first)
```
Hah, does C++ copies Rust's `unwrap_or` with` value_or`, or vice versa?

Only as a remark: In any case not sure this is all that readable, took me a second to figure out.

chriseth replied:
I don't understand this.

cameel commented on test/libsolidity/SolidityExecutionFramework.h:
```diff
@@ -65,6 +66,7 @@ class SolidityExecutionFramework: public solidity::test::ExecutionFramework
 
 	bytes multiSourceCompileContract(
 		std::map<std::string, std::string> const& _sources,
+		std::optional<std::string> const& _sourceName = std::nullopt,
```
This is the name of the entry from `_sources` that can contain external sources?
```suggestion
		std::optional<std::string> const& _mainSourceName = std::nullopt,
```

And I think you should assert that `_sourceName` is in `_sources` (if not `nullopt`).

axic commented on test/Common.h:
```diff
@@ -83,4 +83,6 @@ struct CommonOptions: boost::noncopyable
 	static std::unique_ptr<CommonOptions const> m_singleton;
 };
 
+bool isValidSemanticTestPath(boost::filesystem::path const& _testPath);
```
I'd add documentation here:

```suggestion
// Returns true if the path is within `test/libsolidity/semanticTests` and it does not start with an underscore.
bool isValidSemanticTestPath(boost::filesystem::path const& _testPath);
```

cameel replied:
Yep :) https://github.com/ethereum/solidity/pull/11110#discussion_r616971108

cameel commented on test/tools/isoltest.cpp:
```diff
@@ -26,6 +26,7 @@
 #include <test/EVMHost.h>
 
 #include <boost/algorithm/string/replace.hpp>
+#include <boost/algorithm/string/predicate.hpp>
```
This seems unused.

axic commented on test/libsolidity/semanticTests/externalSource/source_remapping.sol:
```diff
@@ -0,0 +1,15 @@
+==== ExternalSource: ExtSource=_external/external.sol ====
```
Well since remapping were implemented it is fine having them, but why did we wanted them in the first place?

cameel replied:
You suggested it :) https://github.com/ethereum/solidity/issues/10475#issuecomment-737264878
Apparently because OpenZeppelin has lots of relative imports.

Well, I think that we do not really need remappings to support relative imports. They will just work.

But I think these remappings are useful anyway because they will allow us to use sources using direct imports without having to change them. E.g.:

```solidity
# _openzeppelin/contracts/utils.sol
import "@openzeppelin/contracts/array.sol";
library Utils {}

#  _openzeppelin/contracts/array.sol
library Array {}
```
```solidity
==== ExternalSource: @openzeppelin/contracts/array.sol=_openzeppelin/contracts/array.sol ====
==== ExternalSource: _openzeppelin/contracts/utils.sol ====
==== Source: A ====
import "_openzeppelin/contracts/utils.sol";
```

Actually for convenience (if we want to add whole big libraries like OZ) it might be better even to support remapping whole directories:
```solidity
# _openzeppelin/contracts/utils.sol
import "@openzeppelin/contracts/array.sol";
library Utils {}

#  _openzeppelin/contracts/array.sol
library Array {}
```
```solidity
==== ExternalSource: @openzeppelin/=_openzeppelin/ ====
==== Source: A ====
import "@openzeppelin/contracts/utils.sol";
```

cameel commented on test/libsolidity/semanticTests/externalSource/_external/import_with_subdir.sol:
```diff
@@ -0,0 +1 @@
+import "subdir/import.sol";
```
You have test cases using two different naming conventions in the same PR (`import_with_subdir.sol` vs `otherExternal.sol`). I think we should stick to one.

cameel commented on test/TestCaseReader.cpp:
```diff
@@ -137,6 +146,56 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
+				{
+					ErrorList errorList;
+					ErrorReporter errorReporter(errorList);
+					shared_ptr<Scanner> scanner = make_shared<Scanner>(CharStream(externalSourceContent, externalSourceName));
```
Is `externalSourceName` passed to `CharStream` only used in error messages? If so I think it would be more helpful to use `externalSource` here because it's the real path to the file on disk.
```suggestion
					shared_ptr<Scanner> scanner = make_shared<Scanner>(CharStream(externalSourceContent, externalSource));
```

cameel commented on test/TestCaseReader.cpp:
```diff
@@ -137,6 +146,56 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
+							);
+							sources[(externalSourceParentDir / import->path()).generic_string()] = importedSourceContent;
+							sources[import->path()] = importedSourceContent;
+						}
```
After looking at tests and realizing how the whole mechanism is meant to be used, I think this part is unnecessary. There's no need to scan external sources for imports at all. By doing this you're simulating the source discovery performed by `FileReader` and that only adds complexity.

I think that all the necessary sources should be loaded explicitly instead. I.e. `sources` should contain only:
- the current test case (or the files listed as `Source:` if it's a multi-source test)
- the files listed as `ExternalSource:`

It makes the whole system much simpler. Let's say you have a file `_external/contract.sol`:
```
import "@openzeppelin/contracts/utils.sol";
```

All you have to do to use it is to say which file `@openzeppelin/contracts/utils.sol` points at:

```
==== ExternalSource: @openzeppelin/contracts/utils.sol=_external/utils.sol ====
==== ExternalSource: _external/contract.sol ====
import "_external/contract.sol";
```

This is really similar to how we provide source using Standard JSON. The "remapping" here is closer to how the `urls` key works in `sources` dict in the JSON than to what we call remappings on the CLI.

cameel replied:
This seems to be done now. Great.

cameel commented on test/libsolidity/semanticTests/externalSource/source_import.sol:
```diff
@@ -0,0 +1,22 @@
+==== ExternalSource: _external/import.sol ====
+import "_external/external.sol";
+import "_external/otherExternal.sol";
+contract C {
```
I think it would be better if you had to explicitly tell isoltest to make these files available as external sources with:
```
==== ExternalSource: _external/external.sol ====
==== ExternalSource: _external/otherExternal.sol ====
```
otherwise they should cause an import error.

cameel commented on test/TestCaseReader.cpp:
```diff
@@ -137,6 +146,56 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
 			}
+			else if (boost::algorithm::starts_with(line, externalSourceDelimiterStart) && boost::algorithm::ends_with(line, sourceDelimiterEnd))
+			{
+				string externalSource = boost::trim_copy(line.substr(
+					externalSourceDelimiterStart.size(),
+					line.size() - sourceDelimiterEnd.size() - externalSourceDelimiterStart.size()
+				));
+				string externalSourceName = externalSource;
```
You should report an error when `externalSource` is an empty path.

cameel commented on test/libsolidity/semanticTests/externalSource/source_remapping_absolute.sol:
```diff
@@ -0,0 +1,15 @@
+==== ExternalSource: /ExtSource.sol=_external/external.sol ====
```
Some test ideas. I'm listing only files and the `ExternalSources`/`Sources`/`import`s they should contain for brevity:
1) Relative imports in external sources:
    - **EDIT**: Added missing `dir/` segment in `contract.sol` path.
    - File `semanticTests/externalSource/relative_imports.sol`
        - `ExternalSource: _relative_imports/dir/contract.sol`
        - `ExternalSource: _relative_imports/dir/a.sol`
        - `ExternalSource: _relative_imports/dir/B/b.sol`
        - `ExternalSource: _relative_imports/c.sol`
        - `ExternalSource: _relative_imports/D/d.sol`
        - `ExternalSource: _relative_imports/dir/G/g.sol`
        - `ExternalSource: _relative_imports/h.sol`
        - `import "_relative_imports/dir/contract.sol"`
    - File `semanticTests/externalSource/_relative_imports/dir/contract.sol`
        - `import "./a.sol" as a`
        - `import "./B/b.sol" as b`
        - `import "../c.sol" as c`
        - `import "../D/d.sol" as d`
        - `import "./E/../F/../G/./g.sol" as g`
        - `import "../../../../_relative_imports/h.sol" as h`
    - File `semanticTests/externalSource/_relative_imports/dir/a.sol`
    - File `semanticTests/externalSource/_relative_imports/dir/B/b.sol`
        - `import "../c.sol"`
    - File `semanticTests/externalSource/_relative_imports/c.sol`
    - File `semanticTests/externalSource/_relative_imports/D/d.sol`
    - File `semanticTests/externalSource/_relative_imports/dir/G/g.sol`
        - `import "../B/b.sol"`
    - File `semanticTests/externalSource/_relative_imports/h.sol`
2) Duplicate external sources
    - File `semanticTests/externalSource/_duplicate_external_sources.sol`
        - `ExternalSource: _duplicate_external_sources/contract.sol`
        - `ExternalSource: _duplicate_external_sources/contract.sol`
        - `ExternalSource: contract.sol=_duplicate_external_sources/contract.sol`
        - `ExternalSource: contract.sol=_duplicate_external_sources/contract.sol`
        - `Source: _duplicate_external_sources/contract.sol`
        - `Source: contract.sol`
        - `import "_duplicate_external_sources/contract.sol" as contract1`
        - `import "contract.sol" as contract2`
    - File `semanticTests/externalSource/_duplicate_external_sources/contract.sol`
    - Files should contain something that will make it clear whether it's the `Source` or `ExternalSource` that ends up being imported in this situation.
3) External sources with non-normalized paths
    - File `semanticTests/externalSource/_non_normalized_paths.sol`
        - `ExternalSource: _non_normalized_paths/a.sol`
        - `ExternalSource: _non_normalized_paths//a.sol`
        - `ExternalSource: _non_normalized_paths//b.sol`
        - `ExternalSource: C/c.sol=_non_normalized_paths/c1.sol`
        - `ExternalSource: C//c.sol=_non_normalized_paths/c2.sol`
        - `ExternalSource: D//d.sol=_non_normalized_paths/d.sol`
        - `import "_non_normalized_paths//a.sol"`
        - `import "_non_normalized_paths//b.sol"`
        - `import "C//c.sol"`
        - `import "D//d.sol"`
    - File `semanticTests/externalSource/_non_normalized_paths/a.sol`
    - File `semanticTests/externalSource/_non_normalized_paths/b.sol`
    - File `semanticTests/externalSource/_non_normalized_paths/c1.sol`
    - File `semanticTests/externalSource/_non_normalized_paths/c2.sol`
    - File `semanticTests/externalSource/_non_normalized_paths/d.sol`
    - Files should contain something that will make it clear which of these files end up being imported.
4) Remappings with multiple `=` signs:
    - File `semanticTests/externalSource/multiple_equals_signs.sol`
        - `ExternalSource: a=_multiple_equals_signs/contract.sol=c=d.sol`
        - `import "a"`
    - File `semanticTests/externalSource/_multiple_equals_signs/contract.sol=c=d.sol`
5) External files importing files from a multi-source file:
    - File `semanticTests/externalSource/_importing_from_multisource.sol`
        - `ExternalSource: _importing_from_multisource/contract.sol`
        - `Source: a.sol`
        - `import "_importing_from_multisource/contract.sol"`
    - File `semanticTests/externalSource/_importing_from_multisource/contract.sol`
        - `import "a.sol"`
6) External source names containing `./` and `../`
    - File `semanticTests/externalSource/_source_name_starting_with_dots.sol`
        - `ExternalSource: _source_name_starting_with_dots/dir/contract.sol`
        - `ExternalSource: _source_name_starting_with_dots/dir/a.sol`
        - `ExternalSource: _source_name_starting_with_dots/b.sol`
        - `ExternalSource: ./a.sol=_source_name_starting_with_dots/dot_a.sol`
        - `ExternalSource: ../b.sol=_source_name_starting_with_dots/dot_dot_b.sol`
        - `Source: ./a.sol`
        - `Source: ../b.sol`
        - `import "_source_name_starting_with_dots/contract.sol"`
    - File `semanticTests/externalSource/_source_name_starting_with_dots/dir/contract.sol`
        - `import "./a.sol"`
        - `import "../b.sol"`
    - File `semanticTests/externalSource/_source_name_starting_with_dots/dir/a.sol`
    - File `semanticTests/externalSource/_source_name_starting_with_dots/b.sol`
    - File `semanticTests/externalSource/_source_name_starting_with_dots/dot_a.sol`
    - File `semanticTests/externalSource/_source_name_starting_with_dots/dot_dot_b.sol`
    - Files should contain something that will make it clear which of these files end up being imported.

cameel reviewed (commented):
Finally managed to get through the full PR so this is all from me for now.

I have a lot of small suggestions, but these should be quick and easy. There are only a few bigger points:
- I think that parsing files for imports is unnecessary. Instead all importable files should be declared as `Source:` or `ExternalSource:`. This will work very nicely with any kind of import, even relative ones. See https://github.com/ethereum/solidity/pull/11110#discussion_r617047235.
- I have some test suggestions that include corner cases that will likely trip the current implementation (but should work fine with what I'm proposing above).
- There is no need to restrict imports to relative ones. Only actual filesystem paths should be relative, not import paths.
- I pointed out several cases that should in my opinion be handled by raising an exception (rather than assertion or nothing).

cameel commented:
One more thing. I think we should be creating `CompilerStack` in tests without file reader callback (or an empty one that just returns an error. The logic in the default file reader is only suited for command-line usage. It will end up importing the wrong files in tests. If we disable it, we will get much more predictable behavior because only compiler's virtual filesystem will be used.

aarlt commented:
> One more thing. I think we should be creating `CompilerStack` in tests without file reader callback (or an empty one that just returns an error. The logic in the default file reader is only suited for command-line usage. It will end up importing the wrong files in tests. If we disable it, we will get much more predictable behavior because only compiler's virtual filesystem will be used.

I think we use right now an "empty" reader callback for the tests.

`explicit CompilerStack(ReadCallback::Callback _readFile = ReadCallback::Callback());`.

aarlt commented:
@cameel thx a lot for your detailed review! I removed the parser part - it is really not needed, because `ExternalSource` could be used to import the files manually. I just added a commit that reflects your last review. I was not yet able to implement all of the tests that you mentioned, so far only 1) relative imports in external sources.

However, additionally to the things you mentioned in your review, I implemented now the correct logic for updating expectations that are used in tests using external sources. I would continue adding your ideas of the other tests tomorrow.

cameel commented:
> I think we use right now an "empty" reader callback for the tests.

Right. But only if you're not passing any arguments to `CallStack` constructor. I that the case?

cameel commented on test/CommonSyntaxTest.h:
```diff
@@ -81,7 +82,7 @@ class CommonSyntaxTest: public frontend::test::EVMVersionRestrictedTestCase
 	static std::vector<SyntaxTestError> parseExpectations(std::istream& _stream);
 
-	std::map<std::string, std::string> m_sources;
+	solidity::frontend::test::SourceMap m_sources;
```
```suggestion
	frontend::test::SourceMap m_sources;
```

cameel commented on test/TestCaseReader.cpp:
```diff
@@ -137,6 +144,38 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
+				{
+					externalSourceName = boost::trim_copy(externalSourceTarget.substr(0, remappingPos));
+					externalSourceTarget = boost::trim_copy(externalSourceTarget.substr(remappingPos + 1));
+				} else
```
```suggestion
				}
				else
```

cameel commented on test/TestCaseReader.cpp:
```diff
@@ -137,6 +144,38 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
+				fs::path testCaseParentDir = canonical(fs::path{m_fileName}).parent_path();
+				if (!fs::path(externalSourceTarget).is_relative())
+					BOOST_THROW_EXCEPTION(runtime_error("External Source paths need to be relative to the location of the test case."));
+				fs::path externalSourceFullPath = testCaseParentDir / fs::path(externalSourceTarget);
```
```suggestion
				fs::path externalSourceFullPath = testCaseParentDir / externalSourceTarget;
```

aarlt replied:
Hmm.. not sure about this. I think it is safer to "parse" the path using `fs::path` -`externalSourceTarget` is just a `string`.

cameel replied:
I think it does exactly the same processing. If you do something like `path / "a/b/c"` it correctly treats it as 3 subdirs, not as one subdir with a strange name. I use that form often so I don't think an explicit conversion is really needed here.

cameel replied:
Also, I think you could declare `externalSourceTarget` as `fs::path` anyway (as I already mentioned in one of the past comments).

cameel commented on test/TestCaseReader.h:
```diff
@@ -31,6 +32,7 @@ namespace solidity::frontend::test
 struct SourceMap
 {
 	std::map<std::string, std::string> sources;
+	std::vector<std::pair<std::string, std::string>> externalSources;
```
When looking at this at first I thought you're storing the source in both and wanted to comment that you could just store the target for external sources. But I see that's what you're doing. It would be much more apparent if you used `boost::filesystem::path` here:
```suggestion
	std::vector<std::pair<std::string, boost::filesystem::path>> externalSources;
```

cameel commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -368,42 +368,62 @@ void SemanticTest::printSource(ostream& _stream, string const& _linePrefix, bool
+	set<string> externals;
+	for (auto const& [name, path]: m_sources.externalSources)
+	{
+		externals.insert(name.empty() ? path : name);
```
The name cannot be empty in a valid test, can it?

cameel commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -368,42 +368,62 @@ void SemanticTest::printSource(ostream& _stream, string const& _linePrefix, bool
+	}
+
+	for (auto const& [name, source]: m_sources.sources)
+		if (externals.find(name) == externals.end())
```
What if the same name is used for a normal source and an external source? I think that this condition would be fine if there was a check somewhere preventing such a situation.

aarlt replied:
I added an additional check in `test/TestCaseReader.cpp`: 

```
if (sources.count(externalSourceName))
    BOOST_THROW_EXCEPTION(runtime_error("Multiple definitions of test source \"" + externalSourceName + "\"."));
```

cameel replied:
Are `Sources` always loaded before all `ExternalSources`? If not then you might need the same check in the code that parses `Sources`.

aarlt replied:
Ah, sure. This code was already there :)

cameel commented on test/libsolidity/semanticTests/externalSource/_relative_imports/contract.sol:
```diff
@@ -0,0 +1,26 @@
+import "./dir/a.sol" as a;
+import "./dir/B/b.sol" as b;
+import "./c.sol" as c;
+import "./D/d.sol" as d;
+import "./E/../F/../G/./g.sol" as g;
+import "../../../../_relative_imports/h.sol" as h;
```
I'd move `contract.sol` and `G/g.sol` to `dir/` and change the imports to:

```suggestion
import "./a.sol" as a;
import "./B/b.sol" as b;
import "../c.sol" as c;
import "../D/d.sol" as d;
import "./E/../F/../G/./g.sol" as g;
import "../../../../_relative_imports/h.sol" as h;
```

This way you would be able to test the `../`. This is actually how I intended this to work but looks like I made a mistake in my test comment. Sorry for that :)

cameel commented on test/libsolidity/semanticTests/externalSource/_relative_imports/dir/B/b.sol:
```diff
@@ -0,0 +1,11 @@
+import "../../c.sol";
+
+contract B {
+    C _c;
+    constructor() {
+        _c= new C();
```
```suggestion
import "../../c.sol";

contract B {
    C _c;
    constructor() {
        _c = new C();
```

cameel commented on test/libsolidity/semanticTests/externalSource/_relative_imports/dir/B/b.sol:
```diff
@@ -0,0 +1,11 @@
+import "../../c.sol";
```
If this import works then there's something wrong. `c.sol` is one directory up from `b.sol` not two.

```suggestion
import "../c.sol";
```

cameel reviewed (commented):
Here's another round. The code is pretty good now.

I think the biggest remaining thing is to ensure that all the test cases I listed in https://github.com/ethereum/solidity/pull/11110#discussion_r617074805 work correctly. I think that duplicates are not handled correctly yet. Be we should really just disallow duplicate names or external source names matching source names.

aarlt commented:
> > I think we use right now an "empty" reader callback for the tests.
> 
> Right. But only if you're not passing any arguments to `CallStack` constructor. I that the case?

Yes, looks like. `SolidityExecutionFramework` just default constructs `solidity::frontend::CompilerStack`.

chriseth commented on test/TestCaseReader.cpp:
```diff
@@ -137,6 +144,43 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
+					externalSourceName = externalSourceString;
+
+				solAssert(!externalSourceName.empty(), "");
+				externalSourceName = weakly_canonical(fs::path{externalSourceName}).generic_string();
```
Do we really need all this?

aarlt replied:
I only thought that it would be nicer to add this in canonical form, I guess its not really needed.

cameel replied:
Good point. We do not want canonical on source unit names. They're not necessarily paths.

chriseth commented on test/TestCaseReader.cpp:
```diff
@@ -156,7 +200,7 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
 	// Register the last source as the main one
 	sources[currentSourceName] = currentSource;
-	return {{move(sources), move(currentSourceName)}, lineNumber};
+	return {{move(sources), move(externalSources), move(currentSourceName)}, lineNumber};
```
Why do we have to distinguish between sources and external sources?

aarlt replied:
I distinguish sources and external sources to enable the correct generation of expectation updates.

cameel replied:
And sources only have source unit name while for external sources we also need to know the path on disk.

aarlt replied:
Also `sources` contains external sources. It holds also the content of external sources. I didn't wanted to touch the compilation logic of present in `SolidityExecutionFramework`. `sources` and `externalSources` use the same keys, where the value of `source` points to it's content. For `externalSources` it points to the path of that external source. This information is only used to update expectations.

chriseth commented on test/libsolidity/semanticTests/externalSource/_external/external.sol=sol:
```diff
@@ -0,0 +1,5 @@
+contract External {
```
Why the `=` in the file name?

aarlt replied:
This is just the test-case `test/libsolidity/semanticTests/externalSource/multiple_equals_signs.sol`. The test just checks the behaviour if multiple equal signs where used  in a remapping. The first equal sign is used for the remapping. The second one gets correctly detected as a part of the filename.

cameel commented on test/TestCaseReader.h:
```diff
@@ -31,6 +32,7 @@ namespace solidity::frontend::test
 struct SourceMap
 {
 	std::map<std::string, std::string> sources;
+	std::vector<std::pair<std::string, boost::filesystem::path>> externalSources;
```
```suggestion
	std::vector<std::tuple<std::string, boost::filesystem::path>> externalSources;
```
Why not a map though?

aarlt replied:
I thought that it may be important to preserve the original `ExternalSource` order. So that when `isoltest` is updating the expectations, we will have the original order again.

cameel replied:
Well, I think the order might affect which error messages you get or the AST IDs but I think it should not change the execution results. I agree that it might not hurt to keep the order just in case but we already ignore it in `sources` so I'd suggest ignoring it here too for simplicity.

If you ignore it and use a map, it will be easier to check if you do not have duplicates.

aarlt replied:
Ok, I changed that to a map. But I think we don't need to check `externalSources` for duplicates here, because the `source` map also contains external sources.

cameel commented on test/TestCaseReader.cpp:
```diff
@@ -137,6 +144,43 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
+				if (sources.count(externalSourceName))
+					BOOST_THROW_EXCEPTION(runtime_error("Multiple definitions of test source \"" + externalSourceName + "\"."));
+
+				externalSources.emplace_back(externalSourceName, externalSourceTarget);
```
This does not prevent duplicates. I think we should not allow multiple different external sources under the same name.

aarlt replied:
`sources` also contains external sources. Thats why it also prevents duplicates in `externalSources`.

cameel replied:
Ah, ok. I'd assert that then. that would make it obvious.

cameel commented on test/libsolidity/semanticTests/externalSource/non_normalized_paths.sol:
```diff
@@ -0,0 +1,28 @@
+==== ExternalSource: _non_normalized_paths/a.sol ====
+==== ExternalSource: _non_normalized_paths//b.sol ====
+==== ExternalSource: C/////c.sol=_non_normalized_paths/c.sol ====
+==== ExternalSource: C/../////D/d.sol=_non_normalized_paths///d.sol ====
+import "_non_normalized_paths/a.sol";
+import "_non_normalized_paths/b.sol";
+import "C/c.sol";
+import "D/d.sol";
```
The point of the test I suggested was that if you have `ExternalSource: a///b.sol` and `ExternalSource: a/b.sol` then `import "a///b.sol"` should import the one with `///`. That's how it currently works in `CompilerStack` - import paths are not normalized in anyway and I think we want this to behave in tests like that too.

cameel commented on test/libsolidity/semanticTests/externalSource/source_name_starting_with_dots.sol:
```diff
@@ -0,0 +1,22 @@
+==== ExternalSource: /Users/alex/git/solidity/cmake-build-debug/test/tools/a.sol=_source_name_starting_with_dots/dot_a.sol ====
+==== ExternalSource: /Users/alex/git/solidity/cmake-build-debug/test/b.sol=_source_name_starting_with_dots/dot_dot_b.sol ====
```
Why `/Users/alex/git/solidity/cmake-build-debug/`? I don't see any test importing a path like that.

aarlt replied:
oh wow! I think that was introduced by an expectation update. Need to check why that happened.

cameel replied:
Ah, so that's how it got here :)
I suspect that it's because of `weakly_canonical()`. It converts the path to an absolute one if it can.

cameel commented on test/libsolidity/semanticTests/externalSource/source_name_starting_with_dots.sol:
```diff
@@ -0,0 +1,22 @@
+==== ExternalSource: /Users/alex/git/solidity/cmake-build-debug/test/tools/a.sol=_source_name_starting_with_dots/dot_a.sol ====
+==== ExternalSource: /Users/alex/git/solidity/cmake-build-debug/test/b.sol=_source_name_starting_with_dots/dot_dot_b.sol ====
+==== ExternalSource: _source_name_starting_with_dots/dir/a.sol ====
+==== ExternalSource: _source_name_starting_with_dots/b.sol ====
+==== ExternalSource: _source_name_starting_with_dots/dir/contract.sol ====
+import "_source_name_starting_with_dots/dir/contract.sol";
```
What about remappings like `ExternalSource: ./a.sol=_source_name_starting_with_dots/dot_a.sol`? I think we should have a test checking how they work. Such a remapping **should not** match `import "./a.sol"`, That's because the `./` part gets replaced by the path to the importing module when `CompilerStack` evaluates the import.

aarlt replied:
Exactly. It looks like that `import "./a.sol"` would become `import "a.sol"`, that is not found. But within `externalSource/_source_name_starting_with_dots/dir/contract.sol`, we do `import "./a.sol";` and `import "../b.sol";`, that becomes `import "dir/a.sol";` and `import "b.sol";` So I think what you described is already tested, not sure what you mean. E.g. if `import "./a.sol";` would import the defined `_source_name_starting_with_dots/dot_a.sol`, `contract A` would not be found.

cameel replied:
Yeah, it's being tested because the `./a.sol=` remapping is present now. When I commented it was not there. I think it was because of https://github.com/ethereum/solidity/pull/11110#discussion_r618667060.

cameel commented on test/libsolidity/semanticTests/externalSource/_source_name_starting_with_dots/dir/contract.sol:
```diff
@@ -0,0 +1,2 @@
+import "./a.sol";
+import "../b.sol";
```
To reduce the number of files I think you could just put this directly in `_source_name_starting_with_dots.sol`.

aarlt replied:
This is not working, because, if I understood correctly, `./` and `../` will be replaced with a value in relation to the path of the file that is doing these imports. That means if we put this directly in `source_name_starting_with_dots.sol`,  `import "./a.sol";` would become `import "a.sol";`  and  `import "../b.sol";` would become `import "b.sol";`. We didn't defined these files.

cameel replied:
Ah, you're right, sorry :) It has to be in `contract.sol`.

cameel commented on test/libsolidity/semanticTests/externalSource/non_normalized_paths.sol:
```diff
@@ -0,0 +1,28 @@
+==== ExternalSource: _non_normalized_paths/a.sol ====
+==== ExternalSource: _non_normalized_paths//b.sol ====
+==== ExternalSource: C/////c.sol=_non_normalized_paths/c.sol ====
+==== ExternalSource: C/../////D/d.sol=_non_normalized_paths///d.sol ====
+import "_non_normalized_paths/a.sol";
+import "_non_normalized_paths/b.sol";
+import "C/c.sol";
+import "D/d.sol";
+
+contract Contract {
+    A _a;
+    B _b;
+    C _c;
+    D _d;
+    constructor() {
+        _a = new A();
+        _b = new B();
+        _c = new C();
+        _d = new D();
+    }
+    function foo() public returns (uint) {
+        return _a.foo() + _b.foo() + _c.foo() + _d.foo();
+    }
+}
+// ====
+// compileViaYul: also
+// ----
+// foo() -> 0x04f5
```
As discussed on the channel, these tests turned out to be quite verbose and we should try to trim them down a bit. Especially for the ones I suggested in https://github.com/ethereum/solidity/pull/11110#discussion_r617074805 the only relevant things are the paths and imports. You could safely remove almost all functions and these tests will provide the same value. As long as it compiles, it's good so the semantic part can be minimal.

For example in this one I would just use an empty `contract Contract {}`.

In `a.sol` I would only leave `contract A {}` and then import it with
```solidity
import {A} from "_non_normalized_paths//a.sol";
```
This is enough to make the compiler fail if it imports the wrong file and `A` is not available in it.

aarlt replied:
Ok, I simplified this test a bit.

cameel replied:
Looks good. Could you simplify `relative_imports.sol`, `source_name_starting_with_dots.sol`, `_relative_imports/dir/contract.sol` and other files from the `_` dirs that way too? I think there's still a lot that can be cut.

cameel commented on test/libsolidity/semanticTests/externalSource/source_remapping.sol:
```diff
@@ -0,0 +1,15 @@
+==== ExternalSource: ExtSource=_external/external.sol ====
```
I would combine `source_remapping.sol` and `source_remapping_absolute.sol` into a single test.

Also, I would use empty contracts inside. We're only testing paths here and the code inside can be minimal.

cameel reviewed (commented):
I see that you changed the reader to normalize the external source names. I think the previous version was correct. The point is that `CompilerStack` does not normalize them so this mechanism should not either.

Also, I gave some suggestions about trimming down the tests. This was discussed on the channel today. @chriseth pointed out that there are too many tests but I think there's just a bit too much code inside them. They can be just as effective with just imports and directives inside.

cameel commented on test/TestCaseReader.cpp:
```diff
@@ -137,6 +144,41 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
+					externalSourceName = externalSourceString;
+
+				solAssert(!externalSourceName.empty(), "");
+				fs::path externalSourceTarget(fs::path{externalSourceString});
```
You don't need to construct a `path` object to be able to pass it to the `path` constructor :)
```suggestion
				fs::path externalSourceTarget(externalSourceString);
```

cameel commented on test/TestCaseReader.cpp:
```diff
@@ -137,6 +144,41 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
+				solAssert(!externalSourceName.empty(), "");
+				fs::path externalSourceTarget(fs::path{externalSourceString});
+				fs::path testCaseParentDir = m_fileName.parent_path();
+				if (!fs::path(externalSourceTarget).is_relative())
```
```suggestion
				if (!externalSourceTarget.is_relative())
```

cameel commented on test/TestCaseReader.cpp:
```diff
@@ -137,6 +144,41 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
+				fs::path externalSourceFullPath = testCaseParentDir / externalSourceTarget;
+				string externalSourceContent;
+				if (!fs::exists(externalSourceFullPath))
+					BOOST_THROW_EXCEPTION(runtime_error("External Source '" + externalSourceTarget.generic_string() + "' not found."));
```
I think it's fine to have a system-specific path format here:
```suggestion
					BOOST_THROW_EXCEPTION(runtime_error("External Source '" + externalSourceTarget.string() + "' not found."));
```

cameel reviewed (commented):
The PR looks pretty much ready. All the functionality and tests are there, and they all pass.

I only have a couple of minor style comments. I could accept it already but @chriseth wanted the tests to be leaner and I think that they could be trimmed a bit more. They are fine in terms of what they test but they could still be doing the same job with a bit less code.

cameel commented on test/libsolidity/semanticTests/externalSource/_source_name_starting_with_dots/dot_a.sol:
```diff
@@ -0,0 +1,2 @@
+contract Dot_A {
```
```suggestion
contract DotA {
```

cameel commented on test/libsolidity/semanticTests/externalSource/_source_name_starting_with_dots/dot_dot_b.sol:
```diff
@@ -0,0 +1,2 @@
+contract Dot_Dot_B {
```
```suggestion
contract DotDotB {
```

cameel reviewed (approved):
I think this is ready to merge.

Should probably get a final look from @mijovic or @chriseth but as for me, I think there's nothing substantial left to do here.

There are some potential modifications that could make it easier to use with big external libraries, e.g. remapping whole dirs or enabling `FileReader` so that files can be imported without having to list them all but I'm not sure we actually want them and if we do, they're better  done as extensions, in separate PRs.

chriseth commented:
I didn't review it, it's our own code, so easy to change later on.
